File: PathPathJointChildrenCollection.cs

/*****************************************************************************
 *                                                                           *
 *  PathPathJointChildrenCollection.cs                                       *
 *  1 December 2006                                                          *
 *  Project: Crystalline                                                     *
 *  Written by: Richard Sartor                                               *
 *  Copyright © 2006 Metaphysics Industries, Inc.                            *
 *                                                                           *
 *  Converted from C++ to C# on 10 November 2007                             *
 *                                                                           *
 *  An unordered collection of PathJoint objects.                            *
 *                                                                           *
 *****************************************************************************/

using System;
using System.Collections.Generic;
using MetaphysicsIndustries.Collections;

namespace MetaphysicsIndustries.Crystalline
{
	public class PathPathJointChildrenCollection : IList<PathJoint>
	{
        private PathPathJointChildrenCollection() { }
		public PathPathJointChildrenCollection(Path path)
		{
			_parent = path;
		}

		public  void Dispose()
		{
			Clear();

            //_list.Clear();
			//list.Dispose();
            //_list = null;
		}

		public virtual void Add(PathJoint pj)
		{
			if (!this.Contains(pj))
			{
				pj.ParentPath = null;
				_list.Add(pj);
				pj.ParentPath = _parent;
			}
		}

		public virtual void Clear()
		{
			List<PathJoint>	list;
			list = new List<PathJoint>();
			foreach (PathJoint pj in this)
			{
				list.Add(pj);
			}
			foreach (PathJoint pj in list)
			{
				this.Remove(pj);
			}
			list.Clear();
		}

		public virtual bool Contains(PathJoint pj)
		{
			return _list.Contains(pj);
		}

		public virtual void Insert(int i, PathJoint pj)
		{
			_list.Insert(i, pj);
		}

		public virtual IEnumerator<PathJoint> GetEnumerator()
		{
			return _list.GetEnumerator();
		}

		public virtual bool Remove(PathJoint pj)
		{
			bool	b;
			b = _list.Remove(pj);
			if (b && pj.ParentPath == _parent)
			{
				pj.ParentPath = null;
			}
			return b;
		}

		public virtual void CopyTo(PathJoint[] r, int i)
		{
			_list.CopyTo(r, i);
		}

		public virtual void RemoveAt(int i)
		{
			_list.RemoveAt(i);
		}

		public virtual int IndexOf(PathJoint pj)
		{
			return _list.IndexOf(pj);
		}

		public virtual PathJoint Last
		{
			get
			{
				return this[Count - 1];
			}
		}

		public virtual PathJoint this [ int index ]
		{
			get
			{
                return _list[index];
			}
			set
			{
                _list[index] = value;
			}
		}

		public virtual bool IsReadOnly
		{
			get
			{
				return _list.IsReadOnly;
			}
		}

		public virtual int Count
		{
			get
			{
				return _list.Count;
			}
		}

		public virtual PathJoint First
		{
			get
			{
				return this[0];
			}
		}

		  System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
		{
            return GetEnumerator();
		}

        private Path _parent;
        private IList<PathJoint> _list = new List<PathJoint>();
	}
}

File: Crystalline/PathPathJointChildrenCollection.cpp

//////////// source file ////////////


/*****************************************************************************
 *                                                                           *
 *  PathPathJointChildrenCollection.cpp                                      *
 *  1 December 2006                                                          *
 *  Project: Crystalline                                                     *
 *  Written by: Richard Sartor                                               *
 *  Copyright © 2006 Metaphysics Industries                                  *
 *                                                                           *
 *  An unordered collection of PathJoint objects.                            *
 *                                                                           *
 *****************************************************************************/



#include "stdafx.h"
#include "PathPathJointChildrenCollection.h"
#include "Path.h"
#include "PathJoint.h"



NAMESPACE_START;


PathPathJointChildrenCollection::PathPathJointChildrenCollection(Path^ path)
{
	this->parent = path;
	this->list = gcnew List<PathJoint^>;
}

PathPathJointChildrenCollection::~PathPathJointChildrenCollection(void)
{
	this->Clear();
	delete this->list;
}

void PathPathJointChildrenCollection::Add(PathJoint^ pj)
{
	if (!this->Contains(pj))
	{
		pj->ParentPath = nullptr;
		this->list->Add(pj);
		pj->ParentPath = this->parent;
	}
}

bool PathPathJointChildrenCollection::Contains(PathJoint^ pj)
{
	return this->list->Contains(pj);
}

bool PathPathJointChildrenCollection::Remove(PathJoint^ pj)
{
	bool	b;

	b = this->list->Remove(pj);
	if (b && pj->ParentPath == this->parent)
	{
		pj->ParentPath = nullptr;
	}

	return b;
}

void PathPathJointChildrenCollection::Clear(void)
{
	List<PathJoint^>^	list;

	list = gcnew List<PathJoint^>;
	for each (PathJoint^ pj in this)
	{
		list->Add(pj);
	}

	for each (PathJoint^ pj in list)
	{
		this->Remove(pj);
	}

	this->list->Clear();
}

void PathPathJointChildrenCollection::CopyTo(array<PathJoint^>^ r, int i)
{
	this->list->CopyTo(r, i);
}

IEnumerator<PathJoint^>^ PathPathJointChildrenCollection::GetEnumerator(void)
{
	return this->list->GetEnumerator();
}

int PathPathJointChildrenCollection::Count::get(void)
{
	return this->list->Count;
}

bool PathPathJointChildrenCollection::IsReadOnly::get(void)
{
	return this->list->IsReadOnly;
}

System::Collections::IEnumerator^ PathPathJointChildrenCollection::GetEnumerator2(void)
{
	return this->GetEnumerator();
}

int PathPathJointChildrenCollection::IndexOf(PathJoint^ pj)
{
	return this->list->IndexOf(pj);
}

void PathPathJointChildrenCollection::Insert(int i, PathJoint^ pj)
{
	this->list->Insert(i, pj);
}

void PathPathJointChildrenCollection::RemoveAt(int i)
{
	this->list->RemoveAt(i);
}

PathJoint^ PathPathJointChildrenCollection::default::get(int i)
{
	return this->list[i];
}

void PathPathJointChildrenCollection::default::set(int i, PathJoint^ pj)
{
	this->list[i] = pj;
}

PathJoint^ PathPathJointChildrenCollection::First::get(void)
{
	return this[0];
}
PathJoint^ PathPathJointChildrenCollection::Last::get(void)
{
	return this[this->Count - 1];
}



NAMESPACE_END;



